### PR TITLE
Remove `VarInt` and use `ReadExt` and `WriteExt` trait methods instead 

### LINKED ISSUE #1016: VarInt in this project is CompactSize in Bitcoin Core
The Bitcoin Code source code uses the name "CompactSize" for what is named "VarInt" in this project.

"CompactSize/VarInt" in [this project](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/src/consensus/encode.rs#L379) and [in Bitcoin Core](https://github.com/bitcoin/bitcoin/blob/master/src/serialize.h#L246).

In Bitcoin Core, there is also a different integer encoding named [VarInt](https://github.com/bitcoin/bitcoin/blob/master/src/serialize.h#L370). 

I assumed `VarInt` was the same integer encoding for both projects and this confused me for a while.

### PR DESCRIPTION
**Pull request by tcharding**
This the meat and potatoes out of Steven's work in #2133 and also closes #1016

### CODE REVIEW

**Comment by tcharding** [comment]
CC @stevenroose, I picked up your work man, hope you don't mind.

**Comment by tcharding** [comment]
Might close #1016

**Comment by tcharding** [comment]
Name might be wrong, converting to draft to work it out. ref : https://github.com/rust-bitcoin/rust-bitcoin/issues/1016#issuecomment-2193827980

File: bitcoin/src/blockdata/witness.rs [hunk 1]
```diff
@@ -225,8 +225,7 @@ impl Encodable for Witness {
         let content_with_indices_len = self.content.len();
```
Comment (Kixunil) [inline bitcoin/src/blockdata/witness.rs:219]: The 9fe420d2210bfcfab4d0913c9dd94f633496f95a commit message makes it sound like getting the length is impossible without returning it. That's a complete BS.

Instead the message should be something like "Returning `usize` is more convenient for consumers that need to add up the length of multiple encoded items."

File: bitcoin/src/consensus/encode.rs [hunk 2]
```diff
@@ -438,96 +509,21 @@ impl_int_encodable!(i16, read_i16, emit_i16);
...
-                }
-            }
-            n => Ok(VarInt::from(n)),
-        }
-    }
+/// Returns 1 for 0..=0xFC, 3 for 0xFD..=(2^16-1), 5 for 0x10000..=(2^32-1), and 9 otherwise.
+#[inline]
+pub fn varint_size(v: impl IntoU64) -> usize {
+    varint_size_const(v.into_u64())
 }
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply bitcoin/src/consensus/encode.rs:527]: You can't make trait methods `const`.

File: bitcoin/src/consensus/encode.rs [hunk 3]
```diff
@@ -186,6 +186,21 @@ pub fn deserialize_partial<T: Decodable>(data: &[u8]) -> Result<(T, usize), Erro
...
+    /// Convert into a [u64].
+    fn into_u64(self) -> u64;
+}
+
+impl IntoU64 for u8 { fn into_u64(self) -> u64 { self as u64 } }
+impl IntoU64 for u16 { fn into_u64(self) -> u64 { self as u64 } }
+impl IntoU64 for u32 { fn into_u64(self) -> u64 { self as u64 } }
+impl IntoU64 for u64 { fn into_u64(self) -> u64 { self } }
+// NB that we don't support architectures other than 32-bit and 64-bit
+impl IntoU64 for usize { fn into_u64(self) -> u64 { self as u64 } }
```
[the remark this answers is not in the document]
Comment (tcharding) [reply bitcoin/src/consensus/encode.rs:202]: Ugh, sloppy work by me.

File: bitcoin/src/consensus/encode.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply bitcoin/src/consensus/encode.rs:527]: I mean make this `fn` const by being non-generic and use methods for convenience instead.

File: bitcoin/src/consensus/encode.rs [hunk 4]
```diff
@@ -438,96 +509,21 @@ impl_int_encodable!(i16, read_i16, emit_i16);
...
-            }
-            _ => {
-                w.emit_u8(0xFF)?;
-                self.0.consensus_encode(w)?;
-                Ok(9)
-            }
-        }
+/// Returns 1 for 0..=0xFC, 3 for 0xFD..=(2^16-1), 5 for 0x10000..=(2^32-1), and 9 otherwise.
+#[inline]
+pub const fn varint_size_const(v: u64) -> usize {
```
[the remark this answers is not in the document]
Comment (tcharding) [reply bitcoin/src/consensus/encode.rs:498]: Done.

File: bitcoin/src/consensus/encode.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tcharding) [reply bitcoin/src/consensus/encode.rs:511]: Flagging that I did not action this.

**Comment by tcharding** [comment]
Rebased, added a new final patch, and re-wrote PR description.

**Comment by tcharding** [comment]
Perhaps we should instead use `get_size_of_comact_size` to mirror `GetSizeOfCompactSize` from Core

ref: https://github.com/bitcoin/bitcoin/blob/37cdb5f2483cd57077497925caf69a82b0d39471/src/serialize.h#L295C31-L295C35

**Comment by Kixunil** [comment]
@tcharding I'd drop the `get_` prefix, it's not used in Rust much.

Also I think we shouldn't remove `VarInt` but deprecate it first.

**Comment by tcharding** [comment]
`encoded_size_of_compact_size`?

**Comment by tcharding** [comment]
Draft to see what happens with `push_decod` and also #3259

**Comment by stevenroose** [comment]
Obvious concept ACK

**Comment by apoelstra** [comment]
Needs rebase

**Comment by tcharding** [comment]
Rebased, then got lost fixing my emacs, not sure if I touched anything other than rebase.

**Comment by apoelstra** [comment]
Will ACK. FWIW I'm happy with this API and would be willing to pull it out of `internals` (though I guess it'd most sensibly live with `ReadExt`/`WriteExt` which also don't have homes yet).

**Review by apoelstra — DISMISSED** [review]
ACK 1599db253574a97191ee63146ce57040845800e1 successfully ran local tests; nice, though sad that we need a separate non-generic `compact_size_const` method

**Comment by tcharding** [comment]
> though sad that we need a separate non-generic `compact_size_const` method

Agreed.

**Comment by tcharding** [comment]
Re-based, during which `arbitrary` import moved locations, I think that is why the range-diff shows changed lines.

**Comment by tcharding** [comment]
Needs #3395

**Review by apoelstra — APPROVED** [review]
ACK b58dd27c42d56df8a3e05ff719022e240ea0ff81 successfully ran local tests; nice, though sad that we need a separate non-generic `compact_size_const` method

**Comment by tcharding** [comment]
lol, amusingly, your re-post of the 'sad that we ...' confused me, I was sure I'd read that before :)

**Comment by apoelstra** [comment]
Yeah :) I'm not sure what the best way for me to "refresh an ACK" with a nontrivial message is. I want to repeat the message so it shows up in the commit message, but it's weird to read for people on the PR.

**Comment by tcharding** [comment]
Oh, that is ok, now I know it won't surprise me.

**Review by apoelstra — APPROVED** [review]
ACK 18d8b0e469720485ef7cfb9647e3ba66a044fa59 successfully ran local tests

**Comment by tcharding** [comment]
I rekon we can merge this, all @Kixunil's concerns from last month have been either addressed or commented on.

**Pull request merged**